sentry_sdk/integrations/falcon.py

```diff
@@ -0,0 +1,162 @@
+from typing import Any, Callable, Dict
+
+import falcon
+import falcon.api_helpers
+import sentry_sdk.integrations
+from sentry_sdk.hub import Hub
+from sentry_sdk.integrations._wsgi_common import RequestExtractor
+from sentry_sdk.integrations.wsgi import SentryWsgiMiddleware
+from sentry_sdk.utils import capture_internal_exceptions, event_from_exception
+
+
+class FalconRequestExtractor(RequestExtractor):
+
+    def env(self):
+        return self.request.env
+
+    def cookies(self):
+        return self.request.cookies
+
+    def raw_data(self):
+        # As request data can only read once we won't make this available
+        # to Sentry.
+        # TODO(jmagnusson): Figure out if there's a way to support this
+        return None
+
+    def form(self):
+        return None  # No such concept in Falcon
+
+    def files(self):
+        return None  # No such concept in Falcon
+
+    def json(self):
+        # We don't touch falcon.Request.media as that can raise an exception
+        # on non-JSON requests.
+        return self.request._media
+
+
+class SentryFalconMiddleware(object):
+    """Captures exceptions in Falcon requests and send to Sentry"""
+
+    def process_request(self, req, resp, *args, **kwargs):
+        hub = Hub.current
+        integration = hub.get_integration(FalconIntegration)
+        if integration is None:
+            return
+
+        with hub.configure_scope() as scope:
+            scope.add_event_processor(
+                _make_request_event_processor(req, integration)
+            )
+
+
+class FalconIntegration(sentry_sdk.integrations.Integration):
+    identifier = "falcon"
+
+    transaction_style = None
+
+    def __init__(self, transaction_style="uri_template"):
+        # type: (str) -> None
+        TRANSACTION_STYLE_VALUES = ("uri_template", "path")
+        if transaction_style not in TRANSACTION_STYLE_VALUES:
+            raise ValueError(
+                "Invalid value for transaction_style: %s (must be in %s)"
+                % (transaction_style, TRANSACTION_STYLE_VALUES)
+            )
+        self.transaction_style = transaction_style
+
+    @staticmethod
+    def setup_once():
+        # type: () -> None
+        _patch_wsgi_app()
+        _patch_handle_exception()
+        _patch_prepare_middleware()
+
+
+def _patch_wsgi_app():
+    original_wsgi_app = falcon.API.__call__
+
+    def sentry_patched_wsgi_app(self, env, start_response):
+        hub = Hub.current
+        integration = hub.get_integration(FalconIntegration)
+        if integration is None:
+            return original_wsgi_app(self, env, start_response)
+
+        sentry_wrapped = SentryWsgiMiddleware(
+            lambda envi, start_resp: original_wsgi_app(self, envi, start_resp)
+        )
+        with hub.push_scope() as scope:
```

Comment:
You do not have to push a scope here as the WSGI middleware already does this (it makes a new hub which is close enough.

Setting `scope._name` is not strictly necessary, but you can do this in the falcon middleware.

Reply:
Fixed